File: src/game/TurnManager.ts
import type { PlayerId, TurnResult } from '../types';
import { GAME_CONFIG } from './constants';

export class TurnManager {
  currentPlayer: PlayerId = 1;
  bounceCount = 0;
  isLaunched = false;
  turnActive = false;
  leftBase = false;
  pathPointCount = 0;

  startTurn(): void {
    this.bounceCount = 0;
    this.isLaunched = false;
    this.turnActive = true;
    this.leftBase = false;
    this.pathPointCount = 0;
  }

  onLaunch(): void {
    this.isLaunched = true;
    this.leftBase = true;
  }

  onBounce(): void {
    this.bounceCount++;
  }

  canStillBounce(): boolean {
    return this.bounceCount < GAME_CONFIG.maxBounces;
  }

  recordPathPoint(): void {
    this.pathPointCount++;
  }

  switchPlayer(): void {
    this.currentPlayer = this.currentPlayer === 1 ? 2 : 1;
    this.turnActive = false;
  }

  evaluateTurn(
    onOwnTerritory: boolean,
    onOpponentBase: boolean,
    outOfBounds: boolean,
    isStopped: boolean,
  ): TurnResult {
    if (!this.isLaunched) return 'playing';

    if (outOfBounds) return 'failed_out';
    if (onOpponentBase && this.leftBase) return 'failed_penalty';

    if (onOwnTerritory && this.leftBase && isStopped) {
      if (this.bounceCount <= GAME_CONFIG.maxBounces) {
        return 'claimed';
      }
      return 'failed_bounces';
    }

    if (isStopped && this.leftBase) {
      if (this.bounceCount > GAME_CONFIG.maxBounces) return 'failed_bounces';
      if (!onOwnTerritory) return 'failed_bounces';
    }

    return 'playing';
  }
}

File: flutter_app/lib/game/turn_manager.dart
import 'constants.dart';
import 'models.dart';

class TurnManager {
  PlayerId currentPlayer = PlayerId.p1;
  int bounceCount = 0;
  bool isLaunched = false;
  bool turnActive = false;
  bool leftBase = false;

  void startTurn() {
    bounceCount = 0;
    isLaunched = false;
    turnActive = true;
    leftBase = false;
  }

  void onLaunch() {
    isLaunched = true;
    leftBase = true;
  }

  void onBounce() {
    bounceCount++;
  }

  bool canStillBounce() => bounceCount < GameConfig.maxBounces;

  void switchPlayer() {
    currentPlayer = currentPlayer == PlayerId.p1 ? PlayerId.p2 : PlayerId.p1;
    turnActive = false;
  }

  TurnResult evaluateTurn({
    required bool onOwnTerritory,
    required bool onOpponentBase,
    required bool outOfBounds,
    required bool isStopped,
  }) {
    if (!isLaunched) return TurnResult.playing;
    if (outOfBounds) return TurnResult.failedOut;
    if (onOpponentBase && leftBase) return TurnResult.failedPenalty;

    if (onOwnTerritory && leftBase && isStopped) {
      return bounceCount <= GameConfig.maxBounces
          ? TurnResult.claimed
          : TurnResult.failedBounces;
    }

    if (isStopped && leftBase) {
      if (bounceCount > GameConfig.maxBounces) return TurnResult.failedBounces;
      if (!onOwnTerritory) return TurnResult.failedBounces;
    }

    return TurnResult.playing;
  }
}
